astropy_helpers/openmp_helpers.py

```diff
@@ -0,0 +1,133 @@
+#   list of conditions and the following disclaimer.
+# * Redistributions in binary form must reproduce the above copyright notice, this
+#   list of conditions and the following disclaimer in the documentation and/or
+#   other materials provided with the distribution.
+# * Neither the name of the Astropy Team nor the names of its contributors may be
+#   used to endorse or promote products derived from this software without
+#   specific prior written permission.
+#
+# THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND
+# ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
+# WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+# DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE FOR
+# ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
+# (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
+# LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON
+# ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+# (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
+# SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+
+from __future__ import absolute_import, print_function
+
+import os
+import glob
+import tempfile
+import subprocess
+
+from distutils import log
+from distutils.ccompiler import new_compiler
+from distutils.sysconfig import customize_compiler
+from distutils.errors import CompileError, LinkError
+
+from .setup_helpers import get_compiler_option
+
+__all__ = ['add_openmp_flags_if_available']
+
+
+def add_openmp_flags_if_available(extension):
+    """
+    Add OpenMP compilation flags, if available (if not a warning will be
+    printed to the console and no flags will be added)
+
+    Returns `True` if the flags were added, `False` otherwise.
+    """
+
+    ccompiler = new_compiler()
+    customize_compiler(ccompiler)
+
+    tmp_dir = tempfile.mkdtemp()
+
+    CCODE = """
+    #include <omp.h>
+    #include <stdio.h>
+    int main() {
+    #pragma omp parallel
+    printf("nthreads=%d\\n", omp_get_num_threads());
+    }
+    """
+
+    start_dir = os.path.abspath('.')
+
+    if get_compiler_option() == 'msvc':
+        compile_flag = '-openmp'
+        link_flag = ''
+    else:
+        compile_flag = '-fopenmp'
+        link_flag = '-fopenmp'
+
+    try:
+
+        os.chdir(tmp_dir)
+
+        with open('test_openmp.c', 'w') as f:
+            f.write(CCODE)
+
+        os.mkdir('objects')
+
+        # Compile, link, and run test program
+        ccompiler.compile(['test_openmp.c'], output_dir='objects', extra_postargs=[compile_flag])
+        ccompiler.link_executable(glob.glob(os.path.join('objects', '*')), 'test_openmp', extra_postargs=[link_flag])
+        output = subprocess.check_output('./test_openmp').decode('utf-8').splitlines()
```

Comment:
does this decode need to take ``locale.getdefaultlocale()`` or something like that? Maybe it doesn't matter, but it seems odd to be hard-coded as UTF-8?

Reply:
Done